Objects/genobject.c

```diff
@@ -432,16 +436,18 @@ _gen_throw(PyGenObject *gen, int close_on_genexit,
                 Py_DECREF(yf);
                 goto throw_here;
             }
-            gen->gi_running = 1;
+            PyFrameState state = gen->gi_frame->f_state;
+            gen->gi_frame->f_state = FRAME_EXECUTING;
             ret = PyObject_CallFunctionObjArgs(meth, typ, val, tb, NULL);
-            gen->gi_running = 0;
+            gen->gi_frame->f_state = state;
             Py_DECREF(meth);
         }
         Py_DECREF(yf);
         if (!ret) {
             PyObject *val;
             /* Pop subiterator from stack */
-            ret = *(--gen->gi_frame->f_stacktop);
+            gen->gi_frame->f_stackdepth--;
```

Comment:
Should we have an `assert(gen->gi_frame->f_stackdepth >= 0)` sanity check here before decrementing the stack depth? I don't know that it's necessary here, but since it gets removed with `-O` anyways it won't incur a real cost.

Reply:
Note to self: `assert(gen->gi_frame->f_stackdepth > 0);` was added in https://github.com/python/cpython/pull/20803/commits/a22412c5430b3e604fb2d7f765733692f0da7373.